### PR TITLE
Parameterize CompactForm String for optional SCALE impl

### PR DESCRIPTION
**Pull request by ascjones**
By default, parity-scale-codec does not provide Encode/Decode impls for an owned String.
This is only provided under the "full" feature which is not used by the substrate runtime,
because it should not be used for consensus critical code. So in order for the CompactForm
to be integrated into the substrate runtime, or wherever the "full" feature cannot be used,
then we must parameterize the `String` type so that it can be both an `&'static str` on the
runtime side where it is encoded, and a `String` in client/consuming code where it is decoded.

Note that previously I had done something similar with `OwnedForm` discussed with @Robbepop [here](https://github.com/paritytech/scale-info/pull/19#discussion_r472238289). But this or something similar is now required in case we want to generate the metadata directly inside the substrate runtime where `String`s are not allowed.

### CODE REVIEW

File: src/form.rs [hunk 1]
```diff
@@ -76,9 +76,12 @@ impl Form for MetaForm {
 ///
 /// `type String` is owned in order to enable decoding
 #[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Serialize, Debug)]
-pub enum CompactForm {}
+pub struct CompactForm<S = &'static str>(PhantomData<S>);
```
Comment (dvdplm) [inline src/form.rs:79]: For my education, can you elaborate on the choice of default here? Why is `&'static str` the natural choice here?

**Review by dvdplm — APPROVED** [review]
I can't say I fully understand the code surrounding your change here; the docs are a bit "academic" and I have a hard time following them at times. 
The changes in this PR are straight forward and seems like a no-brainer per se.

File: src/form.rs (hunk 1 above)
Comment (ascjones) [reply src/form.rs:83]: When constructing type infos the strings are always `'static`. Only for deserializing/decoding by consumers do we need owned strings.

**Comment by ascjones** [comment]
I'd like @Robbepop to approve this too before merging, because he disagreed with my previous approach https://github.com/paritytech/scale-info/pull/19#discussion_r472238289

**Comment by ascjones** [comment]
Related: https://stackoverflow.com/questions/65348505/runtime-building-string-not-found-in-this-scope

**Comment by ascjones** [comment]
Merged since other work already based on this

**Pull request merged**